internal/controller/promotions/promotions.go

```diff
@@ -439,23 +433,34 @@ func (r *reconciler) promote(
 		// so we are safe from race conditions and can just update the status
 		// TODO: remove all patching of Stage status out of promo reconciler
 		if err = kubeclient.PatchStatus(ctx, r.kargoClient, stage, func(status *kargoapi.StageStatus) {
-			// control-flow Stage history is maintained in Stage controller.
-			// So we only modify history for normal Stages.
-			// (Technically, we should prevent creating promotion jobs on
-			// control-flow stages in the first place)
 			status.LastPromotion = status.CurrentPromotion
 			status.LastPromotion.Status = newStatus
-
 			if newStatus.Phase == kargoapi.PromotionPhaseSucceeded {
-				// Only push promotion to Stage status history if the promotion succeeded
-				status.Phase = kargoapi.StagePhaseVerifying
-				status.CurrentPromotion = nil
-				if stage.Spec.PromotionMechanisms != nil {
+				// Handle specific things that need to happen on success.
+				// 1. Trigger re-verification for re-promotions.
+				// 3. Otherwise, update the current freight and history.
+				// 2. Update the phase to Verifying and clear the current promotion.
+				if status.CurrentFreight != nil &&
+					status.CurrentFreight.Name == targetFreight.Name {
+					if err = kargoapi.ReverifyStageFreight(
```

Comment:
While this absolutely does the job for now, I can't help but wonder if the Stage reconciler could become more self-determined.

Reply:
Agreed. @jessesuen and I have both wanted for a long time that each reconciler will only ever mutate the objects it reconciles and not touch others. That's a larger bit of refactoring that needs to happen at some point. Right now, my biggest concern is getting the functionality correct and then we can circle back to figure out how to clean things up a bit. I know you an I both have various improvements in the works that are chipping away at these things.